design/design-doc-packages-v2.md

```diff
@@ -0,0 +1,1387 @@
+    kind: Package
+    name: pkg-a
+    uid: d9607e19-f88f-11e6-a518-42010a800195
+spec:
+  compositions: # pkg-a is only a bundle of compositions
+  - veryabstract.binfra.pkga.coolorg.io
+  - veryabstract.cinfra.pkga.coolorg.io
+  - veryabstract.dinfra.pkga.coolorg.io
+  ...
+```
+
+6. Check to see if all dependencies in `PackageRevision` exist (N)
+7. Create a `Package` for each dependency
+
+```yaml
+apiVersion: packages.crossplane.io/v1alpha1
+kind: Package
+metadata:
+  name: registry.upbound.io-crossplane-pkg-b
+spec:
+  package: registry.upbound.io/crossplane/pkg-b:v1.9.15 # latest v1.* version
+```
+
+```yaml
+apiVersion: packages.crossplane.io/v1alpha1
+kind: Package
+metadata:
+  name: registry.upbound.io-crossplane-pkg-c
+spec:
+  package: registry.upbound.io/crossplane/pkg-c:v1.3.5 # latest in range
+```
+
+```yaml
+apiVersion: packages.crossplane.io/v1alpha1
+kind: Package
+metadata:
+  name: registry.upbound.io-crossplane-pkg-d
+spec:
+  package: registry.upbound.io/crossplane/pkg-d:v2.5.22 # latest patch version
+```
+
+8. Return with short wait
+
+*Reconcile 4 - Observe Progress of Dependency Resolution*
+
+1. Observe `pkg-a`
+2. Get `Module` and check if `pkg-a` exists (Y)
+3. Check if `pkg-a` is the a dependency of another package (N)
+4. Check for existing `PackageRevision` owned by `pkg-a` (Y)
+5. Check for dependencies in `Module` required by `PackageRevision` (Y)
+  - Assuming those packages have now gotten to the stage where they have been
+    added to `Module`
+
+```yaml
+apiVersion: packages.crossplane.io/v1alpha1
+kind: Module
+metadata:
+  name: packages
+spec:
+  packages:
+  - name: pkg-a
+    image: registry.upbound.io/crossplane/pkg-a:v0.0.1@abcdef
+    dependencies:
+    - registry.upbound.io/crossplane/pkg-b:v1.*
+    - registry.upbound.io/crossplane/pkg-c:v1.2.3-v1.3.5
+    - registry.upbound.io/crossplane/pkg-d:~v2.5.7
+  - name: registry.upbound.io-crossplane-pkg-b
+    image: registry.upbound.io/crossplane/pkg-b:1.9.15@abcdef
+    dependencies:
+    - registry.upbound.io/crossplane/provider-aws:v10.*
+    - registry.upbound.io/crossplane/provider-gcp:v10.*
+  - name: registry.upbound.io-crossplane-pkg-c
+    image: registry.upbound.io/crossplane/pkg-c:v1.3.5@abcdef
+    dependencies:
+    - registry.upbound.io/crossplane/provider-aws:v10.1.1
+    - registry.upbound.io/crossplane/provider-gcp:v10.*
+  - name: registry.upbound.io-crossplane-pkg-d
+    image: registry.upbound.io/crossplane/pkg-d:2.5.22@abcdef
+    dependencies:
+    - registry.upbound.io/crossplane/provider-aws:v10.1.*
+    - registry.upbound.io/crossplane/provider-gcp:v10.3.*
+  compositions:
+  - veryabstract.binfra.pkga.coolorg.io: pkg-a
+  - veryabstract.cinfra.pkga.coolorg.io: pkg-a
+  - veryabstract.dinfra.pkga.coolorg.io: pkg-a
+  crds: # Each of the dependency packages (b,c,d) just bring a single InfrastructureDefinition
+  - abstract.pkgb.coolorg.io: pkg-b
+  - abstract.pkgc.coolorg.io: pkg-c
+  - abstract.pkgd.coolorg.io: pkg-d
```

Comment:
I notice that `PackageRevision` appears to have `compositions` and `infrastructureDefinitions`, while this type has `compositons` and `crds` (which look like they might be from InfrastructureDefinitions?).

In general I'm wondering whether it's possible for us to be consistent with regard to whether packages have knowledge of only Crossplane specific types (composition, infrapub, infradef, managed resource, provider) or deals strictly in generic concepts (CR, CRD).

Reply:
@negz the reasoning here is that in the `Module` we are primarily concerned with ensuring that we are safe to install a `PackageRevision`. We collapse `InfrastructureDefinition` (and potentially `InfrastructurePublication`) into `crds` because they can conflict with each other. For instance, if `pkg-b` brings an `InfrastructureDefinition` with name `abstract.pkgb.coolorg.io`, we must make sure that not only is there not another `InfrastructureDefinition` with that name, but also that there is not another "plain" CRD with that name (i.e. maybe `pkg-c` installed a CRD for `abstract.pkgb.coolorg.io`). `compositions` must be separate because they are instances of a type (i.e. `Composition`) rather than a new type (i.e. an instance of a CRD). Therefore, you could potentially have a `composition` that was safe to install that had the same name as a CRD controlled by a different package.

Conversely, in the `PackageRevision` we are actually concerned with whether `abstract.pkgb.coolorg.io` is an `InfrastructureDefinition` or a plain `CRD` because we will check for existence and ownership in a different manner. For an `InfrastructureDefinition`, we will likely want to create the `InfrastructureDefinition` with labels, then also ensure that the resulting CRD was labeled correctly (this could also potentially be deferred to the controller in core crossplane that handles the creation of the CRD). For a CRD we want to create that directly.

So for the full breakdown: the `Module` has `crds` and `compositions`, while `PackageRevision` would have `crds`, `compositions`, `infrastructureDefinitions`, `infrastructurePublications`. We could easily include the more granular fields in the `Module` as well, but this just seemed like a simplification.

Reply:
Would it be worth consolidating all definitions under a `crds` or `definitions` array? In fact, we can list the actual CRDs that will be generated by scanning IDs / IPs. This would let us standardize on CRD level and could make it easier to make changes on ID and IPs; for example we're thinking to remove IP and make it a field on ID. If it's all standardize on CRD level, then package manager could figure out what CRD those types will resolve eventually even if IP is removed.

Reply:
> @negz the reasoning here is that

@hasheddan Makes sense. I just came back to do a final review pass of this design and found myself asking this question again. Perhaps this information belongs in the design doc?